### ISSUE TITLE
No provider for NgbActiveModal! => "Components as content"

### ISSUE THREAD

**Issue by sscots** [ng-bootstrap/ng-bootstrap#1755]
### Bug description:
When trying to implement the Modal using a component like the example here "Components as content"
https://ng-bootstrap.github.io/#/components/modal/examples 
It works like so.
However I want to also use that component outside of a modal, but when I try to access that component using its selector I get the error:
"Unhandled Promise rejection: No provider for NgbActiveModal! ; Zone: <root> ; Task: Promise.then ; Value: Error: No provider for NgbActiveModal!"

### Link to minimally-working plunker that reproduces the issue:
https://plnkr.co/edit/hZMD3vWphJEm7dwLxVEU?p=preview
You'll see in the console as soon as the app loads, it will throw the error.

### Version of Angular, ng-bootstrap, and Bootstrap:

Angular: 4.2.4

ng-bootstrap: 1.0.0-beta.1

Bootstrap: 4.0.0-alpha.6

**Comment by pkozlowski-opensource** [comment]
This works as designed - you can't access `NgbActiveModal` outside of the modal context: the `NgbActiveModal` injectable is provided by the modal service.

Not sure what you exact use-case is so hard to suggest anything...

**Comment by pkozlowski-opensource** [comment]
Closing based on https://github.com/ng-bootstrap/ng-bootstrap/issues/1755#issuecomment-322891642

**Comment by sscots** [comment]
So if I want to use a component inside and modal and also in another place outside of a modal, how would I go about doing that?

**Comment by pkozlowski-opensource** [comment]
> So if I want to use a component inside a modal and also outside of a modal, how would I go about doing that?

One option would be to have to have 2 components with common parts extracted (composition or inheritance). This is the same situation as with router where you can't inject router-specific classes outside of the router context. 

The other option is not to inject `NgbActiveModal` but use other means of closing a modal (ex. `@Output`).

Having said all this I would really understand your practical use-case. As soon as you are in the modal context you've got a notation of closing a modal (and hence you are  injecting `NgbActiveModal`). If you are outside of modal what do you do with the functionality to close a modal? 

tl;dr; there are technical means of having common code but it all depends on what you are trying to do _exactly_.

**Comment by leovo2708** [comment]
Really I don't understand why Angular can know it as a service when it's not a provider in NgbModalModule:
https://github.com/ng-bootstrap/ng-bootstrap/blob/master/src/modal/modal.module.ts

I run my custom content modal in development and it works, but when running in testing, I get error "No provider for NgbActiveModal".

**Comment by kevinrjones** [comment]
I'm seeing this error in test code. The component has NgbActiveModal injected, and I have imported  NgbModule.forRoot(). Can I use NgbActiveModal in components that are being tested?

**Comment by pkozlowski-opensource** [comment]
@kevinrjones of course, just provide a mock for `NgbActiveModal` (the real deal is only injected when a given component is used as content in modals). hard to say more without seeing the actual code, you might ask more on StackOverflow.

**Comment by oslanier** [comment]
try import ```NgbActiveModal``` on app.modules

**Comment by sscots** [comment]
@oslanier That works. Thanks!

**Comment by danielqa** [comment]
@oslanier That works. Thanks!²

**Comment by dianjuar** [comment]
@oslanier With that you can get an instance of `NgbActiveModal` but you still not able to close the current modal

**Comment by ZinalMehta** [comment]
> ### Bug description:
> When trying to implement the Modal using a component like the example here "Components as content"
> https://ng-bootstrap.github.io/#/components/modal/examples
> It works like so.
> However I want to also use that component outside of a modal, but when I try to access that component using its selector I get the error:
> "Unhandled Promise rejection: No provider for NgbActiveModal! ; Zone: ; Task: Promise.then ; Value: Error: No provider for NgbActiveModal!"
> 
> ### Link to minimally-working plunker that reproduces the issue:
> https://plnkr.co/edit/hZMD3vWphJEm7dwLxVEU?p=preview
> You'll see in the console as soon as the app loads, it will throw the error.
> 
> ### Version of Angular, ng-bootstrap, and Bootstrap:
> Angular: 4.2.4
> 
> ng-bootstrap: 1.0.0-beta.1
> 
> Bootstrap: 4.0.0-alpha.6

no I wanted to Implement Modal as child component but it gives StaticInjectorError that is No provider for NgbActiveModal!

**Comment by dianjuar** [comment]
I made a simple work around, that worked for my requirements.

> ##  Close the modal in your `onDestroy` of the component that open the modal

**Comment by lrochadev** [comment]
> try to provide `NgbActiveModal` on app.module
> 
> ```
> providers: [
>     NgbActiveModal,
>   ]
> ```

It works fine.

**Comment by tijuhasz** [comment]
When you try to use a component as a modal and also somewhere else included in your page, it's best to create the content-component with the content your need, then insert content-component into a modal component and also use it separately. This is the cleanest and most logical way IMO.

**Comment by castor-dev3** [comment]
> @oslanier With that you can get an instance of `NgbActiveModal` but you still not able to close the current modal.
> 
> As @pkozlowski-opensource said, the service is inject only in the instance of the modal is being opened

This also happened to me.
So i've found a workaround to close the modal
`constructor(
        private dialogRef: MatDialogRef<any>,
    ) { }`
`public close() {
        this.dialogRef.close();
    }`

**Comment by mohammadreza1313** [comment]
@oslanier NgbActiveModal is not a Service,how can we inject it in provider as a Service?

**Comment by rangfu** [comment]
> try to provide `NgbActiveModal` on app.module
> 
> ```
> providers: [
>     NgbActiveModal,
>   ]
> ```

This solution works and should be added to the [Components as Content](https://ng-bootstrap.github.io/#/components/modal/examples#component) section of the official docs, so that people know how to use components inside and outside of a modal (or if this isn't the best solution, perhaps some notes could be added to the docs about what the correct solution is).

My use case was that I have a small "Contact us" form, which sometimes makes sense to display inline on the page and other times within a modal.

**Comment by LouisGheysens** [comment]
> try to provide `NgbActiveModal` on app.module
> 
> ```
> providers: [
>     NgbActiveModal,
>   ]
> ```

Legend

**Comment by 4javier** [comment]
Back to original issue's usecase, if your content component has to be used outside modal context too, you can inject dependency as optional
```ts
@Component({...})
export class ContentComponent {
  constructor(@Optional private activeModal: NgbActiveModal) {}
}
```
This way when used as non-modal content DI will not complain.
Just remember to check for `null` before accessing `activeModal`.

It works for `NgbOffCanvas`.

**Comment by ALODEV-GT** [comment]
thanks

**Comment by playanth95978** [comment]
> Back to original issue's usecase, if your content component has to be used outside modal context too, you can inject dependency as optional
> 
> @Component({...})
> export class ContentComponent {
>   constructor(@Optional() private activeModal: NgbActiveModal) {}
> }
> This way, when it's used as non-modal content DI will not complain. Just remember to check for `null` before accessing `activeModal`.
> 
> It works for `NgbOffCanvas`.

Not working with prefer injection using inject() ...

**Comment by Gianpiero-Errigo** [comment]
> Not working with prefer injection using inject() ...

Did you pass `optional: true` as second argument?

**Comment by playanth95978** [comment]
> > Not working with prefer injection using inject() ...
> 
> Did you pass `optional: true` as second argument?

yes i just discovered this options after i post my comment. My bad aha, it works perfectly =>

  protected activeModal = inject(NgbActiveModal, { optional: true });
